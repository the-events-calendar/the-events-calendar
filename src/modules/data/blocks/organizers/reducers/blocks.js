/**
 * External dependencies
 */
import { combineReducers } from 'redux';

/**
 * Internal dependencies
 */
import * as types from './../types';

export const byId = ( state = {}, action ) => {
	switch ( action.type ) {
		case types.ADD_BLOCK_ORGANIZER:
			return {
				...state,
				[ action.payload.id ]: action.payload.organizer,
			};
		case types.REMOVE_BLOCK_ORGANIZER:
			return Object.keys( state ).reduce( ( newState, id ) => {
				if ( id === action.payload.id ) {
					return newState;
				}

				return {
					...newState,
					[ id ]: state[ id ],
				};
			}, {} );
		default:
			return state;
	}
};

export const allIds = ( state = [], action ) => {
	switch ( action.type ) {
		case types.ADD_BLOCK_ORGANIZER:
			return [ ...state, action.payload.organizer ];
		case types.REMOVE_BLOCK_ORGANIZER:
			return state.filter( ( organizer ) => organizer !== action.payload.organizer );
		default:
			return state;
	}
};

export default combineReducers( {
	byId,
	allIds,
} );
