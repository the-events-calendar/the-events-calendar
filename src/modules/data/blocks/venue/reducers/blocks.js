/**
 * External dependencies
 */
import { combineReducers } from 'redux';
import { uniq } from 'lodash';

/**
 * Internal dependencies
 */
import * as types from './../types';

/**
 * Reducer that handles adding and removing blocks from state.
 *
 * This sets the state.events.blocks.venue.blocks.byId state, which holds all venue blocks indexed by clientId.
 *
 * @since 6.2.0
 * @param {Object} state State object.
 * @param {string} action Action being taken.
 * @returns {{}} The new state.
 */
export const byId = ( state = {}, action ) => {
	switch ( action.type ) {
		case types.ADD_BLOCK_VENUE:
			return {
				...state,
				[ action.payload.id ]: action.payload.venue,
			};
		case types.REMOVE_BLOCK_VENUE:
			return Object.keys( state ).reduce( ( newState, id ) => {
				if ( id === action.payload.id ) {
					return newState;
				}

				return {
					...newState,
					[ id ]: state[ id ],
				};
			}, {} );
		default:
			return state;
	}
};

/**
 * Reducer that handles adding and removing blocks from state.
 *
 * This sets the state.events.blocks.venue.blocks.allIds state, which is a numerically indexed array of all venues.
 *
 * @since 6.2.0
 * @param {Object} state State object.
 * @param {string} action Action being taken.
 * @returns {{}} The new state.
 */
export const allIds = ( state = [], action ) => {
	switch ( action.type ) {
		case types.ADD_BLOCK_VENUE:
			return uniq( [ ...state, action.payload.venue ] );
		case types.REMOVE_BLOCK_VENUE:
			return state.filter( venue => venue !== action.payload.venue );
		default:
			return state;
	}
};

/**
 * Reducer that handles setting the showMap and showMapLink values in state.
 *
 * This sets the state.events.blocks.venue.blocks.core state.
 *
 * @since 6.2.0
 * @param {Object} state State object.
 * @param {string} action Action being taken.
 * @returns {{}} The new state.
 */
export const core = ( state = {}, action ) => {
	switch ( action.type ) {
		case types.SET_VENUE_MAP:
			return {
				...state,
				showMap: action.payload.showMap,
			};
		case types.SET_VENUE_MAP_LINK:
			return {
				...state,
				showMapLink: action.payload.showMapLink,
			};
		default:
			return state;
	}
};

export default combineReducers( {
	core,
	byId,
	allIds,
} );
