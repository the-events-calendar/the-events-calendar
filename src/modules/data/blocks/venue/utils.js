/**
 * External dependencies
 */
import { isEmpty, get } from 'lodash';
import { getStateName } from '@moderntribe/events/editor/utils/geo-data';
import { editorDefaults } from '@moderntribe/common/utils/globals';

export const getAddress = ( details = {} ) => {
	const { meta = {} } = details;

	if ( isEmpty( meta ) ) {
		return {};
	}

	return {
		street: get( meta, '_VenueAddress', '' ),
		city: get( meta, '_VenueCity', '' ),
		province: get( meta, '_VenueProvince', '' ),
		zip: get( meta, '_VenueZip', '' ),
		country: get( meta, '_VenueCountry', '' ),
	};
};

export const getCoordinates = ( details = {} ) => {
	const { meta = {} } = details;
	const { _VenueLat = '', _VenueLng = '' } = meta;
	const lat = parseFloat( _VenueLat );
	const lng = parseFloat( _VenueLng );

	return {
		lat: isNaN( lat ) ? null : lat,
		lng: isNaN( lng ) ? null : lng,
	};
};

export const setDefault = ( value, defaultValue ) => value === '' ? defaultValue : value;

/**
 * Get Venue Country
 *
 * @param {Object} meta Object with event meta
 * @returns {string} Venue country
 */
export function getVenueCountry( meta ) {
	let country = get( meta, '_VenueCountry', '' );

	if ( '' === country ) {
		const defaultCountry = editorDefaults().venueCountry;
		const [ countryName ] = defaultCountry || [];
		country = countryName || '';
	}
	return country;
}

/**
 * Get Venue State/Province
 *
 * @param {Object} meta Object of venue meta
 * @returns {string} The venue state or province
 */
export function getVenueStateProvince( meta ) {
	let stateProvince = get( meta, '_VenueStateProvince', '' );

	if ( '' === stateProvince ) {
		const country = getVenueCountry( meta );

		if (
			'US' === country ||
			'United States' === country
		) {
			stateProvince = getStateName( 'US', editorDefaults().venueState );
		} else {
			stateProvince = editorDefaults().venueProvince;
		}
	}
	return stateProvince;
}
