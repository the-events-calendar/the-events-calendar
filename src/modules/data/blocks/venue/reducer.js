/**
 * Internal dependencies
 */
import * as types from './types';
import { editorDefaults, mapsAPI } from '@moderntribe/common/utils/globals';

export const DEFAULT_STATE = {
	venue: editorDefaults().venue ? editorDefaults().venue : 0,
	showMap: mapsAPI().embed,
	showMapLink: mapsAPI().embed,
};

export const defaultStateToMetaMap = {
	venue: '_EventVenueID',
	showMap: '_EventShowMap',
	showMapLink: '_EventShowMapLink',
};

export const setInitialState = ( data ) => {
	const { meta } = data;

	Object.keys( defaultStateToMetaMap ).forEach( ( key ) => {
		const metaKey = defaultStateToMetaMap[ key ];
		if ( meta.hasOwnProperty( metaKey ) ) {
			DEFAULT_STATE[ key ] = meta[ metaKey ];
		}
	} );
};

export default ( state = DEFAULT_STATE, action ) => {
	switch ( action.type ) {
		case types.SET_VENUE:
			return {
				...state,
				venue: action.payload.venue,
			};
		case types.SET_VENUE_MAP:
			return {
				...state,
				showMap: action.payload.showMap,
			};
		case types.SET_VENUE_MAP_LINK:
			return {
				...state,
				showMapLink: action.payload.showMapLink,
			};
		default:
			return state;
	}
};
