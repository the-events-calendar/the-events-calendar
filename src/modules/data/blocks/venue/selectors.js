/**
 * External dependencies
 */
import { get } from 'lodash';
import { createSelector } from 'reselect';
import { mapsAPI } from '@moderntribe/common/utils/globals';

export const getMapEmbed = () => get( mapsAPI(), 'embed', true );
export const venueBlockSelector = ( state ) => state.events.blocks.venue.blocks.core;

export const getVenueByClientId = ( state, props ) => state.events.blocks.venue.blocks.byId[ props.clientId ];

export const getVenuesById = ( state ) => state.events.blocks.venue.blocks.byId;

export const getVenuesInBlock = ( state ) => state.events.blocks.venue.blocks.allIds;

export const getshowMapLink = createSelector( [ venueBlockSelector ], ( block ) => block.showMapLink );

export const getshowMap = createSelector(
	[ venueBlockSelector, getMapEmbed ],
	( block, embed ) => embed && block.showMap
);
