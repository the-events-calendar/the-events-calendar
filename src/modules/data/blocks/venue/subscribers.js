/**
 * External dependencies
 */
import { differenceBy } from 'lodash';

/**
 * Internal dependencies
 */
import { globals } from '@moderntribe/common/utils';
import { store } from '@moderntribe/common/store';
import { actions as formActions } from '@moderntribe/common/data/forms';
import {
	actions as venueActions,
	selectors as venueSelectors,
} from '@moderntribe/events/data/blocks/venue';
import { syncVenuesWithPost } from '@moderntribe/events/blocks/event-venue/data/meta-sync';
import { actions as requestActions } from '@moderntribe/common/store/middlewares/request';

const { getState, dispatch } = store;

/**
 * Returns criteria for comparing blocks.
 *
 * @exports
 * @param {Object} block Object with block attributes and data.
 * @returns {string} Client ID of the block.
 */
export const compareBlocks = block => block.clientId;

/**
 * Checks whether the block is the venue block.
 *
 * @exports
 * @param {Object} block Object with block attributes and data.
 * @returns {boolean} Whether the block is the venue block or not.
 */
export const isVenueBlock = ( block ) => block.name === 'tribe/event-venue';

/**
 * Moves the venue to the trash if appropriate (if it is a draft and was removed).
 *
 * @since 6.2.0
 * @param {number} venue
 */
globals.wpHooks.addAction(
	'tec.events.blocks.venue.maybeRemoveVenue',
	'tec.events.blocks.venue.subscribers',
	( venue ) => {
		const path = `tribe_venue/${ venue }`;
		const options = {
			path,
			actions: {
				success: formActions.deleteEntry( dispatch )( path ),
			},
		};
		dispatch( requestActions.wpRequest( options ) );
	},
);

/**
 * Handles the block that was added.
 *
 * @exports
 * @param {Object} block Object with block attributes and data.
 */
export const handleBlockAdded = ( block ) => {
	// only handle event venue block addition
	if ( ! isVenueBlock( block ) ) {
		return;
	}

	// return early if venue attribute is empty
	if ( ! block.attributes.venue ) {
		return;
	}

	dispatch( venueActions.addVenueInBlock( block.clientId, block.attributes.venue ) );
};

/**
 * Handles the block that was removed.
 *
 * @exports
 * @returns {Function} Function that handles the block that was removed.
 */
export const handleBlockRemoved = () => ( block ) => {
	// only handle event venue block removal
	if ( ! isVenueBlock( block ) ) {
		return;
	}

	const venue = venueSelectors.getVenueByClientId( getState(), block );

	// remove venue from block state
	if ( venue ) {
		dispatch( venueActions.removeVenueInBlock( block.clientId, venue ) );

		/**
		 * Moves the venue to the trash if appropriate (if it is a draft and was removed).
		 *
		 * @since 6.2.0
		 * @param {number} venue
		 */
		globals.wpHooks.doAction( 'tec.events.blocks.venue.maybeRemoveVenue', venue );
	}

	syncVenuesWithPost();
};

/**
 * Handles changes in the blocks in the editor.
 *
 * @exports
 * @param {Array} currBlocks Array of current blocks in the editor.
 * @param {Array} prevBlocks Array of previous blocks in the editor.
 */
export const onBlocksChangeHandler = ( currBlocks, prevBlocks ) => {
	const blocksAdded = differenceBy( currBlocks, prevBlocks, compareBlocks );
	const blocksRemoved = differenceBy( prevBlocks, currBlocks, compareBlocks );

	if ( blocksAdded.length ) {
		blocksAdded.forEach( handleBlockAdded );
	}

	if ( blocksRemoved.length ) {
		blocksRemoved.forEach( handleBlockRemoved( currBlocks ) );
	}
};

/**
 * Listener for blocks change in the editor.
 *
 * @exports
 * @param {Function} selector Selector function to get current blocks.
 * @returns {Function} Listener that subscribes to WP store.
 */
export const onBlocksChangeListener = ( selector ) => {
	let holdBlocks = selector();
	return () => {
		const prevBlocks = holdBlocks;
		const currBlocks = selector();
		holdBlocks = currBlocks;

		if (
			prevBlocks.length !== currBlocks.length ||
			differenceBy( currBlocks, prevBlocks, compareBlocks ).length
		) {
			onBlocksChangeHandler( currBlocks, prevBlocks );
		}
	};
};

/**
 * @function subscribe
 * @description This subscribes to any changes in the wp data store.
 *              Since state and attribute changes should not be called
 *              in `componentDidMount()` and `componentWillUnmount()`,
 *              global state changes (tribe common store) and meta changes
 *              are handled in a global subscriber.
 */
const subscribe = () => {
	globals.wpData.subscribe(
		onBlocksChangeListener(
			globals.wpDataSelectCoreEditor().getBlocks,
		),
	);
};

export default subscribe;
