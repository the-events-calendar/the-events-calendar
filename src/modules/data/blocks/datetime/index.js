/**
 * Internal dependencies
 */
import * as reducer from './reducer';
import * as types from './types';
import * as actions from './actions';
import * as selectors from './selectors';
import sagas from './sagas';

export default reducer.default;
export {
	reducer,
	types,
	actions,
	selectors,
	sagas,
};
