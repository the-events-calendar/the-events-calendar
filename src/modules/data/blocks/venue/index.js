/**
 * Internal dependencies
 */
import * as reducer from './reducer';
import * as selectors from './selectors';
import * as actions from './actions';
import * as types from './types';
import * as utils from './utils';

export default reducer.default;
export { reducer, selectors, actions, types, utils };
