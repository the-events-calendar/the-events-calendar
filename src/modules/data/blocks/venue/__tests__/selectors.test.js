/**
 * Internal dependencies
 */
import { selectors } from '@moderntribe/events/data/blocks/venue';

const state = {
	events: {
		blocks: {
			venue: {
				blocks: {
					allIds: [ 42 ],
					byId: {},
					core: {},
				},
			},
		},
	},
};

jest.mock( '@moderntribe/common/utils/globals', () => ( {
	dateSettings: () => ( {} ),
	editorDefaults: () => ( {
		venue: 0,
		venueCountry: '',
		venueState: '',
		venueProvince: '',
	} ),
	list: () => ( {
		countries: {},
		us_states: {},
	} ),
	mapsAPI: () => ( {
		embed: true,
	} ),
	wpHooks: {
		addAction: jest.fn(),
		addFilter: jest.fn(),
		removeAction: jest.fn(),
		removeFilter: jest.fn(),
		doAction: jest.fn(),
		applyFilters: jest.fn( ( tag, value ) => value ),
		hasAction: jest.fn( () => false ),
		hasFilter: jest.fn( () => false ),
	},
} ) );

describe( '[STORE] - Venue selectors', () => {
	it( 'Should return the map embed config', () => {
		expect( selectors.getMapEmbed() ).toMatchSnapshot();
	} );

	it( 'Should return the venue block', () => {
		expect( selectors.venueBlockSelector( state ) ).toMatchSnapshot();
	} );

	it( 'Should select the showMap', () => {
		expect( selectors.getshowMap( state ) ).toMatchSnapshot();
	} );

	it( 'Should select the showMapLink', () => {
		expect( selectors.getshowMapLink( state ) ).toMatchSnapshot();
	} );
} );
