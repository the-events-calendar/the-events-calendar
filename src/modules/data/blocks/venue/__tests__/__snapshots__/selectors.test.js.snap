// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[STORE] - Venue selectors Should return the map embed config 1`] = `true`;

exports[`[STORE] - Venue selectors Should return the venue block 1`] = `{}`;

exports[`[STORE] - Venue selectors Should select the showMap 1`] = `undefined`;

exports[`[STORE] - Venue selectors Should select the showMapLink 1`] = `undefined`;
